Volunteer RSVP limit; Volunteer waitlist

MissAndry opened the pull request:
This adds a Volunteer RSVP limit and a Volunteer waitlist to events, which should help better manage Volunteer attendance.

tjgrathwell commented on app/controllers/rsvps_controller.rb:
```diff
@@ -27,9 +27,13 @@ def create
     Rsvp.transaction do
-      if @event.at_limit? && @rsvp.role_student?
+      if @event.students_at_limit? && @rsvp.role_student?
         @rsvp.waitlist_position = (@event.rsvps.maximum(:waitlist_position) || 0) + 1
```
I think this needs to find the maximum waitlist position for **student waitlist RSVPs** only, since the waitlist sizes / positions will differ between students and volunteers.

tjgrathwell commented on app/controllers/rsvps_controller.rb:
```diff
@@ -27,9 +27,13 @@ def create
       end
+
+      if @event.volunteers_at_limit? && @rsvp.role_volunteer?
+        @rsvp.waitlist_position = (@event.rsvps.maximum(:waitlist_position) || 0 ) + 1
```
I think this needs to find the maximum waitlist position for **volunteer waitlist RSVPs** only, since the waitlist sizes / positions will differ between students and volunteers.

tjgrathwell commented on app/models/event.rb:
```diff
@@ -72,20 +74,26 @@ def meetup_url meetup_event_id
 
   def validate_rsvp_limit
-    return unless persisted? && student_rsvp_limit
+    return unless persisted? && student_rsvp_limit || persisted? && volunteer_rsvp_limit
```
The duplicate `persisted?` can be factored out here

tjgrathwell commented on app/models/event.rb:
```diff
@@ -72,20 +74,26 @@ def meetup_url meetup_event_id
+    return unless persisted? && student_rsvp_limit || persisted? && volunteer_rsvp_limit
 
-    if student_rsvp_limit < student_rsvps_count
+    if (student_rsvp_limit < student_rsvps_count) || (volunteer_rsvp_limit < volunteer_rsvps_count)
```
These should probably be added as two separate errors (for students & volunteers)

tjgrathwell commented on app/models/event.rb:
```diff
@@ -238,6 +250,39 @@ def checkiner?(user)
+        end
+      end
+
+      Rsvp.transaction do
```
I don't think the inner `Rsvp.transaction do` is needed, because we're already in a transaction

MissAndry commented:
Thanks, Travis! I'll update the code to reflect your comments later this week.

tjgrathwell commented:
Hi @MissAndry, this is pretty cool. Here are the big issues that jump out at me (in addition to the more minor things I've added as line comments):

* Earlier this week (9663b104e34bf3dad836f1d47550f39271a36384) I moved the `reorder_waitlist!` function into a WaitlistManager that now lives in [app/services/waitlist_manager.rb](https://github.com/railsbridge/bridge_troll/blob/master/app/services/waitlist_manager.rb). So your `reorder_waitlist!` amendments should go there.
* Your commits add a bunch of screenshot HTML and PNG files from test failures that should just be deleted. Also `test.sqlite3` shouldn't be checked in either.

Also, I don't think `volunteer_rsvp_limit` should be required for every event: organizers should be allowed to leave it blank (in which case infinite volunteers are allowed). This is especially important so that all the existing events will still be `.valid?`. Setting the `volunteer_rsvp_limit` to a value lower than the currently RSVP'd volunteers afterward should still produce a validation error.

Lastly, could you rebase + squash together the commits to reduce the noisy ones ("merged" etc). Ideally, every commit in the sequence tests the feature code it's adding (and passes the tests).

Thanks!

MissAndry commented:
Done deal. Thanks for the feedback! I'll incorporate that functionality as
soon as I can :)
On Apr 1, 2015 10:59 PM, "Travis Grathwell" <notifications@github.com>
wrote:

> Hi @MissAndry <https://github.com/MissAndry>, this is pretty cool. Here
> are the big issues that jump out at me (in addition to the more minor
> things I've added as line comments):
>
>    - Earlier this week (9663b10
>    <https://github.com/railsbridge/bridge_troll/commit/9663b104e34bf3dad836f1d47550f39271a36384>)
>    I moved the reorder_waitlist! function into a WaitlistManager that now
>    lives in app/services/waitlist_manager.rb
>    <https://github.com/railsbridge/bridge_troll/blob/master/app/services/waitlist_manager.rb>.
>    So your reorder_waitlist! amendments should go there.
>    - Your commits add a bunch of screenshot HTML and PNG files from test
>    failures that should just be deleted. Also test.sqlite3 shouldn't be
>    checked in either.
>
> Also, I don't think volunteer_rsvp_limit should be required for every
> event: organizers should be allowed to leave it blank (in which case
> infinite volunteers are allowed). This is especially important so that all
> the existing events will still be .valid?. Setting the
> volunteer_rsvp_limit to a value lower than the currently RSVP'd
> volunteers afterward should still produce a validation error.
>
> Lastly, could you rebase + squash together the commits to reduce the noisy
> ones ("merged" etc). Ideally, every commit in the sequence tests the
> feature code it's adding (and passes the tests).
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-88719485>
> .
>

tjgrathwell commented:
Hey @MissAndry, have you had any time to work on this?

MissAndry commented:
Yes, but not as much as I would've liked! It should be done soon.

On Mon, Apr 13, 2015 at 1:00 AM, Travis Grathwell <notifications@github.com>
wrote:

> Hey @MissAndry <https://github.com/MissAndry>, have you had any time to
> work on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-92227121>
> .
>

tjgrathwell commented:
How's it going with this @MissAndry? It would be nice to get this feature out of pull-request land and into the world :earth_americas: !

MissAndry commented:
Agreed! I am having some problems with the validations, though. For some
reason a null event keeps getting created
On Apr 23, 2015 9:33 AM, "Travis Grathwell" <notifications@github.com>
wrote:

> How's it going with this @MissAndry <https://github.com/MissAndry>? It
> would be nice to get this feature out of pull-request land and into the
> world [image: :earth_americas:] !
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-95606614>
> .
>

tjgrathwell commented:
If you push up your in-progress code and describe what's going wrong I'd be happy to evaluate it and see if I can track down the problem :smile_cat:

MissAndry commented:
Sure thing!

Currently the problem seems to be that the volunteer_limit is allowed to be
a NaN and that the check I tried to put in place
("#volunteer_limit_allowed") isn't catching cases when volunteer_limit is
nil.

Thank you for your help, I appreciate it

On Sat, Apr 25, 2015 at 12:58 AM, Travis Grathwell <notifications@github.com
> wrote:

> If you push up your in-progress code and describe what's going wrong I'd
> be happy to evaluate it and see if I can track down the problem [image:
> :smile_cat:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-96135272>
> .
>

tjgrathwell commented:
Hey @MissAndry,

The main problem I see with this version is that the `validate_rsvp_limit` method is crashing for the cases where `volunteer_rsvp_limit` is not set. The testrun produces errors like:
```
NoMethodError:
  undefined method `<' for nil:NilClass
# ./app/models/event.rb:104:in `validate_rsvp_limit'
```

which traces back to the line
```
if (student_rsvp_limit < student_rsvps_count) || (volunteer_rsvp_limit < volunteer_rsvps_count)
```
which should maybe be changed to something like
```
if (student_rsvp_limit < student_rsvps_count) || (volunteer_rsvp_limit && (volunteer_rsvp_limit < volunteer_rsvps_count))
```

Additionally, the models/event_spec file is failing a test around this [shoulda_matchers](https://github.com/thoughtbot/shoulda-matchers) expectation:
```
it { should validate_numericality_of(:volunteer_rsvp_limit) }
```

in this way:
```
Failure/Error: it { should validate_numericality_of(:volunteer_rsvp_limit) }
    Expected errors to include "is not a number" when volunteer_rsvp_limit is set to "abcd",
    got no errors
```

This is because that test will by default use a blank event and vary the `volunteer_rsvp_limit` attribute to check for errors. An event will not have `volunteer_rsvp_limit` errors unless `allow_volunteer_limit` is set, so it needs to look something like this instead:
```
context 'when the event has a volunteer rsvp limit' do
  subject { Event.new(allow_volunteer_limit: true) }

  it { should validate_numericality_of(:volunteer_rsvp_limit) }
end
```

**However**, I think it may be better to omit the extra `allow_volunteer_limit` boolean entirely and just rely on the presence/absence of the `volunteer_rsvp_limit` field. (As long as sufficient guidance is given on the event form telling organizers that changing the limit from '10' to blank will promote everyone from the volunteer waitlist.)

One more stylistic thing: in the Event model, for symmetry with the `student_rsvp_limit` validation, the `volunteer_rsvp_limit` validation code should be brought under the `with_options(unless: :historical?)` block. Such that that entire block looks like this:

```
  with_options(unless: :historical?) do |normal_event|
    normal_event.with_options(if: :allow_student_rsvp?) do |workshop_event|
      workshop_event.validates_numericality_of :student_rsvp_limit, only_integer: true, greater_than: 0
      workshop_event.validate :validate_rsvp_limit
    end

    normal_event.with_options(if: :volunteer_limit_allowed?) do |workshop_event|
      workshop_event.validates_numericality_of :volunteer_rsvp_limit, only_integer: true, greater_than: 0
      workshop_event.validate :validate_volunteer_rsvp_limit
    end
  end
```

The `(unless: :historical?)` is ensuring that these validations don't apply to really old (imported from meetup) events. Technically this doesn't matter for `volunteer_rsvp_limit` since it only applies when set, but it feels better to keep those two validations close together. I hope someday to move the `:historical?` stuff out of the Event model and into some other model or mixin.

tjgrathwell commented:
Did you get a chance to work on this? Anything I can help with?

MissAndry commented:
Hey, I have some time to work on this today and I should have a little bit
of time tomorrow & later on in the week - I'll let you know if I need any
help!

On Sun, May 3, 2015 at 9:44 PM, Travis Grathwell <notifications@github.com>
wrote:

> Did you get a chance to work on this? Anything I can help with?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-98573763>
> .
>

MissAndry commented:
Hey, I just pushed this up! Please take a look and let me know what you
think

On Sat, May 9, 2015 at 5:10 PM, Megan Gaebler <megan.gaebler@gmail.com>
wrote:

> Hey, I have some time to work on this today and I should have a little bit
> of time tomorrow & later on in the week - I'll let you know if I need any
> help!
>
> On Sun, May 3, 2015 at 9:44 PM, Travis Grathwell <notifications@github.com
> > wrote:
>
>> Did you get a chance to work on this? Anything I can help with?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-98573763>
>> .
>>
>
>

tjgrathwell commented:
Looks improved, except it seems there are still some lines in `RsvpsController` trying to set the `@rsvp.waitlist_position` to the maximum of all RSVP waitlist positions, see https://github.com/railsbridge/bridge_troll/pull/315/files#diff-8f333f5a407b28ce62f67c1aa85d6fffR38

It looks to me like the last `if` statement,
```
if @event.volunteers_at_limit? && @rsvp.role_volunteer?
  @rsvp.waitlist_position = (@event.rsvps.maximum(:waitlist_position) || 0 ) + 1
end
```
should be deleted.

The `promote_from_vlnteer_waitlist` function should just be `promote_from_volunteer_waitlist` -- don't let an angry computer tell you how long your function names should be.

Additionally those screenshot .html files from earlier test failures are still in the commit logs, and should be scrubbed out before this can be merged.

Once you've got that sorted out, can you use `git rebase` to squash this changeset into one commit? 

You could also rebase your changes against the latest master (I've made some changes to `Event` that your code has trouble auto-merging with) but if you're not feeling up for that I can do the merge manually when accepting this PR.

MissAndry commented:
Awesome, I'll work on this this week. Sorry it's taking so long!

On Tue, May 12, 2015 at 12:22 AM, Travis Grathwell <notifications@github.com
> wrote:

> Looks improved, except it seems there are still some lines in
> RsvpsController trying to set the @rsvp.waitlist_position to the maximum
> of all RSVP waitlist positions, see
> https://github.com/railsbridge/bridge_troll/pull/315/files#diff-8f333f5a407b28ce62f67c1aa85d6fffR38
>
> It looks to me like the last if statement,
>
> if @event.volunteers_at_limit? && @rsvp.role_volunteer?
>   @rsvp.waitlist_position = (@event.rsvps.maximum(:waitlist_position) || 0 ) + 1
> end
>
> should be deleted.
>
> The promote_from_vlnteer_waitlist function should just be
> promote_from_volunteer_waitlist -- don't let an angry computer tell you
> how long your function names should be.
>
> Additionally those screenshot .html files from earlier test failures are
> still in the commit logs, and should be scrubbed out before this can be
> merged.
>
> Once you've got that sorted out, can you use git rebase to squash this
> changeset into one commit?
>
> You could also rebase your changes against the latest master (I've made
> some changes to Event that your code has trouble auto-merging with) but
> if you're not feeling up for that I can do the merge manually when
> accepting this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-101137273>
> .
>

MissAndry commented:
I've pulled from the national master and am now getting an error when I run
the tests: "undefined method `allowed_operating_system_ids' for
#<Event:0x007fa3efe19408>"

It's showing up on pretty much every test that uses an event. I've run "be
rake db:migrate" and imported the new column, but running the tests still
gives the same error. What am I missing?

On Wed, May 13, 2015 at 8:17 PM, Megan Gaebler <megan.gaebler@gmail.com>
wrote:

> Awesome, I'll work on this this week. Sorry it's taking so long!
>
> On Tue, May 12, 2015 at 12:22 AM, Travis Grathwell <
> notifications@github.com> wrote:
>
>> Looks improved, except it seems there are still some lines in
>> RsvpsController trying to set the @rsvp.waitlist_position to the maximum
>> of all RSVP waitlist positions, see
>> https://github.com/railsbridge/bridge_troll/pull/315/files#diff-8f333f5a407b28ce62f67c1aa85d6fffR38
>>
>> It looks to me like the last if statement,
>>
>> if @event.volunteers_at_limit? && @rsvp.role_volunteer?
>>   @rsvp.waitlist_position = (@event.rsvps.maximum(:waitlist_position) || 0 ) + 1
>> end
>>
>> should be deleted.
>>
>> The promote_from_vlnteer_waitlist function should just be
>> promote_from_volunteer_waitlist -- don't let an angry computer tell you
>> how long your function names should be.
>>
>> Additionally those screenshot .html files from earlier test failures are
>> still in the commit logs, and should be scrubbed out before this can be
>> merged.
>>
>> Once you've got that sorted out, can you use git rebase to squash this
>> changeset into one commit?
>>
>> You could also rebase your changes against the latest master (I've made
>> some changes to Event that your code has trouble auto-merging with) but
>> if you're not feeling up for that I can do the merge manually when
>> accepting this PR.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-101137273>
>> .
>>
>
>

lilliealbert commented:
@MissAndry You're probably just missing a `rake db test:prepare` for that new migration. (Rails 4 seems to be a liar about the whole 'automatically loading the test schema' thing, as far as I can tell.)

MissAndry commented:
That was totally it, thanks :)

On Wed, May 13, 2015 at 10:26 PM, Lillie Chilen <notifications@github.com>
wrote:

> @MissAndry <https://github.com/MissAndry> You're probably just missing a rake
> db test:prepare for that new migration. (Rails 4 seems to be a liar about
> the whole 'automatically loading the test schema' thing, as far as I can
> tell.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/railsbridge/bridge_troll/pull/315#issuecomment-101897402>
> .
>

tjgrathwell commented:
A couple more things about your latest code:

I think everywhere you're calling `reorder_student_waitlist!` needs to instead invoke some version of `reorder_waitlist!` that considers students+volunteers together. As-is, it doesn't look like `reorder_volunteer_waitlist!` is ever called, so if a volunteer destroys their RSVP or the organizer increases the volunteer RSVP limit it won't promote other volunteers from the waitlist.

I think your changes to the `.gitignore` can be reverted; `test.sqlite3` is already in the `.gitignore` from the [`db/*` line](https://github.com/railsbridge/bridge_troll/blob/master/.gitignore#L15), and the bug that was causing `screenshot_whatever.html` files to show up in the project root has been fixed on master.

tjgrathwell commented:
Hey @MissAndry, this PR was getting a bit long in the tooth, so I committed a squashed version with some amendments as 22c729a0e4fe2611cf9a31d8beecfe4abfbcd98d. It should be live now and ready to use for future workshops.

Thanks so much for your work on this feature!
